File: practical04/ex4.c
#include <stdio.h>
#include <math.h>

#define N 1300

float degtorad(float arg);
float trapint(int npoints); //trapezoidal integral

float y[N];
float x[N];

int main(void){

    int i;
    //N=sizeof(y)/sizeof(int); //is the length of the y array as size of int is 4 bytes

    for(i=0;i<N;i++){ //initialisation of x in degrees

        x[i]=(60./(N-1))*i;
    }

    for(i=0;i<N;i++){ //conversion of x to radians

        x[i]=degtorad(x[i]);
    }
printf("%f\n",x[N-1]);
    for(i=0;i<N;i++){ //computation of tan

        y[i]=tan(x[i]);
    }

    float integral;
    integral=trapint(N);

    printf("Calucated integral = %f \n", integral);
    printf("Exact solution = %f \n", log(2));
    printf("Difference = %f \n", fabs(integral-log(2)) ); //fabs is the math.h absolute value


return(0);
}


float degtorad(float arg){ //conversion to radians

    return((M_PI*arg)/180.0);
}


float trapint(int npoints){

    float sum;
    int i;
    sum=0;

    sum=y[0]+y[npoints-1]; //endpoints

    for(i=1;i<npoints-1;i++){
    sum = sum + 2*y[i]; //the argument of tan is each equidistant point
    //printf("%f\n",sum);
    }

    sum=(x[npoints-1]-x[0])*sum/(2*npoints); //prefactors

return(sum);
}

File: practical04/ex4.f90
module global !global arrays x and y

integer(4),parameter:: N=1300
real(4)::integral
real(4):: x(N),y(N)

end module global


program ex4
use global

    integer(4)::i
    real(4)::diff
    
    do i=1,N,1
        x(i)=(60./(N-1))*(i-1) ! from 0 to 60 degrees
    end do

    do i=1,N,1
        call degtorad(x(i),x(i))
    end do

    do i=1,N,1
        y(i)=tan(x(i))
        !write(*,*) y(i)
    end do

    call trapint(N,integral)

    diff= abs(log(2.) - integral)

    write(6,'(a,f10.8,/,a,f10.8,/,a,f10.8)') 'calculated integral = ',integral,'exact solution = ', log(2.), 'difference = ',diff

    
end program ex4


subroutine degtorad(arg,arg2)
    real (kind=4), intent(in) :: arg
    real (kind=4), intent(out) :: arg2
    real (kind=4) :: pi=3.1415927
    arg2=(pi*arg)/180.0
end subroutine degtorad


subroutine trapint(npoints,summ)
    use global

    integer(4), intent(in):: npoints
    real(4), intent(out):: summ

    integer(4) :: i

    summ=0.
    summ = summ+y(1)+y(npoints)

    do i=1,npoints-1,1
            summ = summ + 2*y(i)
    end do

    summ=(x(N)-x(1))*summ/(2*npoints)

end subroutine trapint
